### PR TITLE
web_editor: correct cursor at editable root

### PR DESCRIPTION
**Pull request by cammarosano**
[FIX] web_editor: fix selection on editable root
[ADD] web_editor: add unit tests for selection correction
[FIX] web_editor: fix typo that led to unexistant property read

task-3128747
opw-3106752

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 1]
```diff
@@ -3980,6 +3994,89 @@ export class OdooEditor extends EventTarget {
...
+        // Utils
+        const restoreLastSelection = () => {
+            const lastSelection = this._latestComputedSelectionInEditable;
+            if (!lastSelection || lastSelection.anchorNode === this.editable) {
+                selection.removeAllRanges();
+                return;
+            }
+            selection.setBaseAndExtent(lastSelection.anchorNode, lastSelection.anchorOffset,
+                lastSelection.focusNode, lastSelection.focusOffset);
+        }
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4029]: Wouldn't `historySetSelection` serve your purpose here wihout the need for a new local function ?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 2]
```diff
@@ -3980,6 +3994,89 @@ export class OdooEditor extends EventTarget {
...
+            setCursorStart(nodeAfterCursor);
+        } else if (paragraphRelatedElements.includes(nodeBeforeCursor.nodeName)) {
+            // Cursor is right after a p.
+            setCursorEnd(nodeBeforeCursor);
+        } else {
+            // Cursor is between two non-p blocks
+            const p = insertParagraph(nodeAfterCursor, 'before');
+            setCursorStart(p);
+        }
+    }
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4078]: I suggest the following rewrite to get rid of the local function:
```js
if (nodeAfterCursor && paragraphRelatedElements.includes(nodeAfterCursor.nodeName)) {
    // Cursor is right before a p.
    setCursorStart(nodeAfterCursor);
} else if (nodeBeforeCursor && paragraphRelatedElements.includes(nodeBeforeCursor.nodeName)) {
    // Cursor is right after a p.
    setCursorEnd(nodeBeforeCursor);
} else {
    this._recordHistorySelection(true);
    // Prevent saving a history step with selection at root.
    if (this._currentStep.selection.anchorNodeOid === 'root') {
        this._currentStep.selection = {};
    }
    const p = this.document.createElement('p');
    p.append(this.document.createElement('br'));
    if (!nodeAfterCursor) {
        // Cursor is at the end of the editable.
        element.append(p);
    } else if (!nodeBeforeCursor) {
        // Cursor is at the beginning of the editable.
        element.prepend(p);
    } else {
        // Cursor is between two non-p blocks
        element.before(p);
    }
    element[method](p);
    this.historyStep();
    if (this._currentStep.selection.anchorNodeOid === 'root') {
        this._currentStep.selection = {};
    }
    setCursorStart(p);
}
```

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 3]
```diff
@@ -3980,6 +3994,89 @@ export class OdooEditor extends EventTarget {
...
+     * Places the cursor in a safe place (not the editable root).
+     * Inserts an empty paragraph when there's no other way to insert text
+     * before/after a block.
+     * If selection was result of an arrow key press, no new 'P' is inserted.
+     *
+     * @param {Selection} selection
+     * @param {String} currentKeyPress
+     */
+    _fixSelectionOnEditableRoot(selection, currentKeyPress) {
+        if (!selection || selection.anchorNode !== this.editable) return;
```
[the remark this answers is not in the document]
Comment (cammarosano) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4007]: Ok... I was just making it a bit safer for eventual future careless callers.
I've put a JSDoc description on the `selection` parameter instead.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 4]
```diff
@@ -3980,6 +3994,89 @@ export class OdooEditor extends EventTarget {
...
+            nodesToSkip.push('HR');
+        }
+        let nodeAfterCursor = this.editable.childNodes[selection.anchorOffset];
+        let nodeBeforeCursor = nodeAfterCursor && nodeAfterCursor.previousSibling;
+        while (nodeAfterCursor && nodesToSkip.includes(nodeAfterCursor.nodeName)) {
+            nodeAfterCursor = nodeAfterCursor.nextSibling;
+        }
+        while (nodeBeforeCursor && nodesToSkip.includes(nodeBeforeCursor.nodeName)) {
+            nodeBeforeCursor = nodeBeforeCursor.previousSibling;
+        }
```
[the remark this answers is not in the document]
Comment (cammarosano) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4019]: You're right.
`nodeBeforeCursor` is simply `nodeAfterCursor.previousElementSibling` (no loop).
Skipping void elements is only necessary when handling the arrow key presses, so I moved the logic to inside those blocks.
And I used `!node.chilNodes.length`  as a test for "void" element. It's not exactely the same thing, as "having no child nodes" is different than" cannot have child nodes" (definition of void element)... but I think that's true in OdooEditor (There's always at least some `<br>` or ZWS in any of our non-void elements... right??) ?

Alternatively, we have an util called `isVoidElement`. It would handle the 'HR' case, but not the others in this list: https://developer.mozilla.org/en-US/docs/Glossary/Void_element.

[the remark this answers is not in the document]
Comment (cammarosano) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4019]: A possible test for void element would be `!element.outerHTML.includes('</')`

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 2 above)
Comment (cammarosano) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4067]: Good one!

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 5]
```diff
@@ -3996,84 +3996,64 @@ export class OdooEditor extends EventTarget {
...
-        } else if (paragraphRelatedElements.includes(nodeBeforeCursor.nodeName)) {
+        } else if (nodeBeforeCursor && paragraphRelatedElements.includes(nodeBeforeCursor.nodeName)) {
             // Cursor is right after a p.
             setCursorEnd(nodeBeforeCursor);
-        } else {
-            // Cursor is between two non-p blocks
-            const p = insertParagraph(nodeAfterCursor, 'before');
+        // Handle other forms of selection.
+        // Insert a new 'P' is selection resulted from a mouse click.
+        } else if (this._currentMouseState === 'mousedown') {
```
Comment (cammarosano) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4038]: @dmo-odoo , this changes a bit the spec that we had agreed upon.
We assumed that any selection change that was not triggered by a key press was necessarily triggered by a mouse click.
But this is not always true, as historyUndo and historyRedo lead to selection changes.
And sometimes undoing a step puts the selection at the editable root. This could lead to an "un-undoable" paragraph insertion. 
Consider the following editable:
```
<hr>
...
```
If the first thing you do when starting editing is placing the cursor before the `hr`, our newly introduced mechanism will insert a `<p>` before it.
But if the user does not want that, (s)he will hit undo. `historyRevert` will remove the `<p>` and, at this point, put the selection on the editable root (at least on my laptop + Chrome). There's no recorded selection on that step, so `historySetSelection` won't change the selection.
As a result the selection is placed before the `<hr>` again, the  mechanism will kick in again and a new `<p>` will be inserted before the `<hr>`. The undo was undone...

So I propose that the paragraph insertion would only happen if the selection change was the result of a mouse click. Otherwise, the selection would simply be removed. Note that this second scenario is rather rare, so it's more like a fallback for weird corner cases.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: The `isVoidElement` util can be fixed to include the other elements. It used to be an obscure util from the GA which only matched like BR I think, we added HR to it to fix a bug and there is another task to add IMG to it too. Theoretically it should have _all_ void elements in it. My point is: use this util and feel free to adapt it if you think it does not handle all the relevant cases.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 5 above)
Comment (cammarosano) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4038]: > Isn't that precisely the spec that we agreed upon ?

Yes... we (or just I) just did not think of the scenario where a selection change is done by a history undo.

> I'm not sure I understand. What about changing the selection with the arrows ? Would you remove the selection too ?

No. Arrow key presses are handled like this:
```
 // Handle arrow key presses.
        if (currentKeyPress === 'ArrowRight' || currentKeyPress === 'ArrowDown') {
            // Skip void elements.
            while (nodeAfterCursor && !nodeAfterCursor.childNodes.length) {
                nodeAfterCursor = nodeAfterCursor.nextElementSibling;
            }
            if (nodeAfterCursor) {
                setSelection(...getDeepestPosition(nodeAfterCursor, 0));
            } else {
                this.historyResetLatestComputedSelection(true);
            }
```
The "unsafe" nodes are skipped until a "safe" node is found and the cursor it put there. Unless you reach the end of the document: in this case, the selection if put back where it was. The end result is the cursor not moving (so no "poof"/disappearing).
Theoretically, `this.historyResetLatestComputedSelection(true);` could put the cursor at the editable root, and the function would be triggered again, this time making it "poof"/gone!
But in practice, before the user pressed an arrow, the cursor would be in a "legal" position, so I don't see this scenario happening.

TL;DR: arrow key presses are handled in the 2 first `if` statements. The removal of a selection is really a fallback: it is in the `else` block at the end of the function.

Comment (cammarosano) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4038]: What this function does is:
- it puts the cursor on a valid place, being mindful of not undoing an arrow key press movement.
- it inserts an empty paragraph if necessary (no other way to put the cursor between two blocks), but only following a mouse click. No P is inserted by an arrow key press (**1**) or by a history undo (**2**).

We had agreed upon **1** (right?), and now I'm proposing **2**.

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4038]: Ok let's squash those and merge this then, thank you ! 👍

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (cammarosano) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: Adding `|| node.tagName === 'BR'` to ` isVoidElement` simply breaks 63 tests... :cry: 
`isVoidElement` is used by other pieces of code like:
- @ sanitize:
```
           // Ensure elements which should not contain any content are tagged
            // contenteditable=false to avoid any hiccup.
            if (
                isVoidElement(node) &&
                node.getAttribute('contenteditable') !== 'false'
            ) {
                node.setAttribute('contenteditable', 'false');
            }
```
- @ OdooEditor._activateContentEditable:
```
        const editableAreas = this.options.getContentEditableAreas(this).filter(node => !isVoidElement(node));
        for (const node of editableAreas) {
            if (!node.isContentEditable) {
                node.setAttribute('contenteditable', true);
            }
        }
```
We (normally) would not expect to find a BR at the editable root, so I decided to not touch this function, as it covers our 'HR' case as it is.
Agreed?

[the remark this answers is not in the document]
Comment (Zinston) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: @dmo-odoo Right I see... So we have 2 concepts: (1) the element is void, and (2) the element is _not allowed_ to have content. Void elements (like `br` or `img`) are not allowed to have content but that's by original design so the browser won't allow it anyway and we don't need to worry about it. But there are other elements that we have to manage ourselves (like `i.fa`).
image
Is that correct so far?
If so, we could keep `isVoidElement` and add something to the effect of:
```js
function isFakeVoidElement(node) {
    return node.classList.contains('fa') || node.classList.contains('o_fake_void');
function isNotAllowedContent(node) {
    return isVoidElement(node) || isFakeVoidElement(node);
```
